pkg/utils/blobstorage.go

```diff
@@ -0,0 +1,272 @@
+// TODO:
+// We should arrange the logic for blob storage in the pipeline-backend.
+// Now, we use blob storage in worker and service. The logic are close but not the same.
+// We should refactor the logic to make it more compact and easier to maintain for worker and service.
+// This will be addressed in ins-7091
+
+package utils
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"mime"
+	"net/url"
+	"strings"
+
+	"github.com/gabriel-vasile/mimetype"
+	"github.com/gofrs/uuid"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/instill-ai/pipeline-backend/config"
+	"github.com/instill-ai/x/blobstorage"
+
+	artifactpb "github.com/instill-ai/protogen-go/artifact/artifact/v1alpha"
+)
+
+// UploadBlobDataAndReplaceWithURLsParams is the parameters for the UploadBlobDataAndReplaceWithURLs function.
+type UploadBlobDataAndReplaceWithURLsParams struct {
+	// NamespaceID is the namespace ID.
+	NamespaceID string
+	// RequesterUID is the requester UID.
+	RequesterUID uuid.UUID
+	// DataStructs are the data structs to be uploaded.
+	DataStructs []*structpb.Struct
+	// Logger is the logger.
+	Logger *zap.Logger
+	// ArtifactClient is the artifact public service client.
+	ArtifactClient *artifactpb.ArtifactPublicServiceClient
+}
+
+// UploadBlobDataAndReplaceWithURL uploads the unstructured data in the structs to minio and replaces the data with the URL.
+// Before calling this function, ctx should have been set with the request metadata.
+func UploadBlobDataAndReplaceWithURLs(ctx context.Context, params UploadBlobDataAndReplaceWithURLsParams) ([]*structpb.Struct, error) {
+	updatedDataStructs := make([]*structpb.Struct, len(params.DataStructs))
+	for i, dataStruct := range params.DataStructs {
+		updatedDataStruct, err := uploadBlobDataAndReplaceWithURL(ctx, uploadBlobDataAndReplaceWithURLParams{
+			namespaceID:    params.NamespaceID,
+			requesterUID:   params.RequesterUID,
+			dataStruct:     dataStruct,
+			logger:         params.Logger,
+			artifactClient: params.ArtifactClient,
+		})
+		if err != nil {
+			// Note: we don't want to fail the whole process if one of the data structs fails to upload.
```

Comment:
nit: you could use the `logger` you're passing around to signal that something failed but the execution continued.